### ISSUE TITLE
`ArrayExpression` does not include `SpreadElement` as possible value of elements property

### ISSUE THREAD

**Issue by qnighy** [typescript-eslint/typescript-eslint#4963]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-estree

### Playground Link

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQMoFFXKgUx3AXzgDMoIQ4ByAARgE8wcBnAYymDBgFpmAbYAHYwA9AFcYwXk0oBuALAAoRcVECWEiALi8IEMAHkAbjijpeOEDiFMAFAEMoUO3QBcKDFlw4AdAEFHzugAHmC4TEzAmgCUiIpwJNBwNiyaTPA45hQQxHAOTnTeGRZWMEwxCHHxcMLCcAASdgIAJuZwLHZMzNrAANb4KQJpuXAAvHAA2t5TAEYAuvIKVXDAOTZFIN70jKMjYwBEqKE4dk1mxUJ7MdO4dj0LSzVwTRBwTOQ4MAAWggDmCVBwL74HAhMIRTSVAiKKEKIA

### Repro Code

```typescript
import { TSESTree } from '@typescript-eslint/utils';

function loopOverElements(array: TSESTree.ArrayExpression) {
  for (const elem of array.elements) {
    // Handle cases like const a = [...b];
    if (elem.type === "SpreadElement") break;
    // do something for the expression
  }
}
```

### Expected Result

The `elem` variable should have a type of `TSESTree.Expression | TSESTree.SpreadElement` and the code above should typecheck.

### Actual Result

The `elem` variable has a type of `TSESTree.Expression` and the code above doesn't typecheck because the type of `elem.type` doesn't intersect with `"SpreadElement"`.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.23.0` |
| `@typescript-eslint/parser`            | `5.23.0` |
| `@typescript-eslint/scope-manager`     | `5.23.0` |
| `@typescript-eslint/typescript-estree` | `5.23.0` |
| `@typescript-eslint/type-utils`        | `5.23.0` |
| `@typescript-eslint/utils`             | `5.23.0` |
| `TypeScript`                           | `4.6.4` |
| `ESLint`                               | `8.15.0` |
| `node`                                 | `16.10.0` |
